src/GraphQL/Language/AST/SourceLocation.cs

```diff
@@ -3,63 +3,32 @@ namespace GraphQL.Language.AST
     /// <summary>
     /// Provides information regarding the location of a node within a document's original query text.
     /// </summary>
-    public class SourceLocation
+    public readonly struct SourceLocation
     {
         /// <summary>
         /// Initializes a new instance with the specified parameters.
         /// </summary>
-        public SourceLocation(int line, int column, int start = -1, int end = -1)
+        public SourceLocation(int start = -1, int end = -1)
         {
-            Line = line;
-            Column = column;
-            Start = start;
-            End = end;
+            _start = start + 1;
+            _end = end + 1;
         }
 
+        private readonly int _start;
         /// <summary>
         /// Returns the start position within the query text.
         /// </summary>
-        public int Start { get; }
+        public readonly int Start => _start - 1;
 
+        private readonly int _end;
         /// <summary>
         /// Returns the end position within the query text.
         /// </summary>
-        public int End { get; }
+        public int End => _end - 1;
 
         /// <summary>
-        /// Returns the line number within the query text.
+        /// Indicates whether the structure has been initialized with a valid values.
         /// </summary>
-        public int Line { get; }
-
-        /// <summary>
-        /// Returns the column number within the query text.
-        /// </summary>
-        public int Column { get; }
-
-        /// <inheritdoc/>
-        public override string ToString() => $"line={Line}, column={Column}, start={Start}, end={End}";
-
-        /// <summary>
-        /// Compares this instance to another based on the line and column values.
-        /// </summary>
-        protected bool Equals(SourceLocation other)
-        {
-            return Line == other.Line && Column == other.Column;
-        }
-
-        /// <inheritdoc/>
-        public override bool Equals(object obj)
-        {
-            if (obj is null)
-                return false;
-            if (ReferenceEquals(this, obj))
-                return true;
-            if (obj.GetType() != GetType())
-                return false;
-            return Equals((SourceLocation)obj);
-        }
-
-        /// <inheritdoc/>
-        public override int GetHashCode() => (Line, Column).GetHashCode();
+        public bool IsSet => _start >= 0 && _end >= _start;
```

Comment:
We can move this check directly into the one line in validationcontext that uses it, or just assume that SourceLocation is always valid (and remove the +1 / -1 stuff).